### DISCUSSION TITLE
Unset unknown/new pk values when using em.assign

### DISCUSSION THREAD

**Discussion by Langstra** [mikro-orm/mikro-orm#3623] [Ideas]
**Is your feature request related to a problem? Please describe.**
When using `em.assign()` to upsert graph entities it is possible to create or update entities. Updating entities based on pk is very useful and when the pk in the data does not match the pk of the entity a new entity is created. The problem here is that the pk could already be in use, which results in either duplicating the entry in the database or attaching an entity that you might not want to.

**Describe the solution you'd like**
Have a flag `clearUnknownPk` that would unset the pk on the data if the pk is not the in entity that you are updating. Some examples are these.

```ts
const projectEntity = {
  id: 1,
  tasks: [{id: 1, name: 'Do chores'}]
};

const projectData = {
  id: 1,
  tasks: [{id: 1, name: 'Do chores'}, {id: 7, name: 'Random other chore'}]
};

em.assign(projectEntity, projectData);
console.log(projectEntity);
{
  id: 1,
  tasks: [{id: 1, name: 'Do chores'}, {id: 2, name: 'Random other chore'}]
};
```

For entities with composite pk's this would not work I think. Entities with composite keys could be used for many to many relations and to attach some data to that relation. For instance
```ts
const projectEntity = {
  id: 1,
  projectUsers: [{userId: 1, projectId: 1},{userId: 2, projectId: 1}] // note that a different project could now have the same users connected
} 
```

Now when assigning new ProjectUsers unsetting the pk would give a problem I think, since cleaning the userId would result in that no user would be connected to the project.

**Comment by B4nan** [comment]
Let's first discuss what your problem is, this seems a bit weird to me (and the fact that it wouldn't work for more than just a simple PK is a bit of a no-no for me, I don't want half-baked features).

We are talking about the [updateByPrimaryKey](https://mikro-orm.io/docs/entity-helper#updating-deep-entity-graph) flag, right? Isn't that already doing what you want? I don't think we can do much better than that.

I will definitely need to see a better example of the problem you want to solve with this, ideally executable repro - your example is not really valid, the last console.log is showing a DTO, not entity state, and I believe it already works like that (better say, you won't see a difference from a console.log, as it's about hidden entity states). I failed to understand both the problem as well as the proposed fix. Lastly, have you seen the `em.upsert` that currently lives in master? As you mentioned upserting, and `assign` is not really for that, it just modifies the entity state (and if it will insert or update - that is already decided before you call it).

edit: oh now I see the difference (id 2 vs id 7), and this makes me even more confused, how do you expect that to work? :] I mean I understand you'd like a flag to ignore the provided PK (which seems like quite a bad idea to me on its own, very confusing behavior), and issue an insert for the second entity, generating a new PK for it instead of using the `id: 7` provided by the user. But this itself would break another invariant, as you wouldn't be able to use this for entity with explicit (not generated) PK at all or even provide a value for a generated PK.

**Comment by Langstra** [comment]
Maybe my example wasn't that clear after all. I tried to create a test case to show what I mean. Upsert only works for single entities I think? I am trying this with a graph entity. 

Here is, hopefully, a better example. 
https://github.com/RiskChallenger/mikro-orm/commit/d004cebaad6d255ec2bf763636aa41c1159157a8

**Reply by B4nan** [reply]
I understand what you want. What I don't understand is why you would want that, or how come you end up with such constraints.

I can't justify a feature that is about ignoring explicitly provided values from the user. This still sounds like you want to bend the ORM instead of fixing your workflow around it.

**Reply by Langstra** [reply]
I think these constraints are not that strange, but yes they come from our workflow. However, I think it is strange to set the id of an entity yourself when you have a pk that is auto incrementing. 

I think this feature is great for everybody that has an API that allows users to graph upsert entities.
